esmvaltool/diag_scripts/thermodyn_diagtool/lorenz_cycle.py

```diff
@@ -0,0 +1,1080 @@
+    flux.draw(filen, list_lorenz)
+    lec = float(kteout) + float(kseout) + float(kzout)
+    return lec
+
+
+def gauaw(n_y):
+    """Compute the Gaussian coefficients for the Gaussian grid conversion.
+
+    @author: Valerio Lembo
+    """
+    c_c = (1 - (2 / math.pi)**2) / 4
+    eps = 0.00000000000001
+    k_k = n_y / 2
+    p_a = np.zeros(n_y)
+    p_a[0:k_k] = bsslzr(k_k)
+    p_w = np.zeros(n_y)
+    for i_l in range(k_k):
+        x_z = np.cos(p_a[i_l] / math.sqrt((n_y + 0.5)**2 + c_c))
+        iterr = 0.
+        zsp = 1.0
+        while (abs(zsp) > eps and iterr <= 10):
+            pkm1 = x_z
+            pkm2 = 1.0
+            for n_n in range(2, n_y, 1):
+                p_k = ((n_n * 2 - 1.0) * x_z * pkm1 - (n_n - 1.0) * pkm2) / n_n
+                pkm2 = pkm1
+                pkm1 = p_k
+            pkm1 = pkm2
+            pkmrk = (n_y * (pkm1 - x_z * p_k)) / (1.0 - x_z**2)
+            zsp = p_k / pkmrk
+            x_z = x_z - zsp
+            iterr = iterr + 1
+        if iterr > 15:
+            sys.exit("*** no convergence in gauaw ***")
+        p_a[i_l] = x_z
+        p_w[i_l] = (2.0 * (1.0 - x_z**2)) / ((n_y**2) * (pkm1**2))
+        p_a[n_y - 1 - i_l] = -p_a[i_l]
+        p_w[n_y - 1 - i_l] = p_w[i_l]
+    psi = p_a
+    pgw = p_w
+    return psi, pgw
+
+
+def globall_cg(d3v, g_w, d_s, dims):
+    """Compute the global and hemispheric averages.
+
+    @author: Valerio Lembo
+    """
+    nlev = dims[0]
+    nlat = dims[2]
+    ntp = dims[3]
+    gmn = np.zeros([3, ntp - 1])
+    aux1 = np.zeros([nlev, int(nlat / 2), ntp - 1])
+    aux2 = np.zeros([nlev, int(nlat / 2), ntp - 1])
+    aux1v = np.zeros([nlev, ntp - 1])
+    aux2v = np.zeros([nlev, ntp - 1])
+    nhem = int(nlat / 2)
+    fac = 1 / G * PS / 1e5
+    for l_l in range(nlev):
+        for i_h in range(nhem):
+            aux1[l_l, i_h, :] = fac * np.real(d3v[l_l, i_h, :]) * g_w[i_h]
+            aux2[l_l, i_h, :] = (fac * np.real(d3v[l_l, i_h + nhem - 1, :]) *
+                                 g_w[i_h + nhem - 1])
+        aux1v[l_l, :] = (np.nansum(aux1[l_l, :, :], axis=0) / np.nansum(
+            g_w[0:nhem]) * d_s[l_l])
+        aux2v[l_l, :] = (np.nansum(aux2[l_l, :, :], axis=0) / np.nansum(
+            g_w[0:nhem]) * d_s[l_l])
+    gmn[1, :] = (np.nansum(aux1v, axis=0) / np.nansum(d_s))
+    gmn[2, :] = (np.nansum(aux2v, axis=0) / np.nansum(d_s))
+    gmn[0, :] = 0.5 * (gmn[1, :] + gmn[2, :])
+    return gmn
+
+
+def init(logfile, filep):
+    """Ingest input fields as complex fields and initialise tables.
+
+    Receive fields t,u,v,w as input fields in Fourier
+    coefficients  (time,level,wave,lon), with real as even and imaginary parts
+    as odd. Convert them to complex fields for Python.
+
+    Arguments:
+        - filenc: name of the file containing the input fields;
+        - logfile: name of the file containing the table as a .txt file.
+
+    Author:
+    Valerio Lembo, University of Hamburg, 2019
+    """
+    log = open(logfile, 'w')
```

Comment:
Always use `open` in a with statement, as this will ensure the file is closed again, even if something goes wrong.
```python
    with open(logfile, 'w') as log:
        log.write('something')
```